### PR TITLE
Recurrent layers

### PR DESCRIPTION
**Pull request by DeNeutoy**

### CODE REVIEW

File: allennlp/common/tensor.py [hunk 1]
```diff
@@ -1,10 +1,85 @@
...
+        A 2D binary mask of shape (batch_size, sequence_length) to
+        calculate the per-batch sequence lengths from.
+
+    Returns
+    -------
+    A torch.LongTensor of shape (batch_size,) representing the lengths
+    of the sequences in the batch.
+    """
+    # Flip mask to make padded elements equal to one.
+    inverse_mask = mask == 0
```
Comment (matt-gardner) [inline allennlp/common/tensor.py]: Can't you just sum the mask here?  Seems simpler.

File: allennlp/common/tensor.py [hunk 2]
```diff
@@ -1,10 +1,85 @@
...
+def get_dropout_mask(dropout_probability: float, shape: List[int]):
+    """
+    Parameters
+    ----------
+    dropout_probability: float, Probability of dropping a dimension of the input.
+    shape: Shape of the tensor you are generating a mask for.
+
+    Return
+    ------
+    A torch.FloatTensor consisting of the binary mask scaled by 1/ (1 - dropout_probability).
```
Comment (matt-gardner) [inline allennlp/common/tensor.py]: "the binary mask"?  Can you give some more detail here?  It's not obvious what computation this method is performing from this docstring.  Maybe an initial sentence at the beginning of the docstring that says something like "Computes an element-wise dropout mask, where each element in a tensor with shape ``shape`` is dropped out with probability ``dropout_probability``."

File: allennlp/common/tensor.py [hunk 3]
```diff
@@ -1,10 +1,85 @@
...
+    (this scaling ensures expected values and variances of the output of applying this mask
```
Comment (matt-gardner) [inline allennlp/common/tensor.py]: Remove the parentheses here, and just make this a normal sentence.

File: allennlp/modules/augmented_lstm.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
+from typing import Optional, Tuple
+import torch
```
Comment (matt-gardner) [inline allennlp/modules/augmented_lstm.py]: Blank line separating built in from third party imports.

File: allennlp/modules/augmented_lstm.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    An LSTM with Recurrent Dropout and the option to use highway
+    connections between layers.
+
+    Parameters
+    ----------
+    input_size : int, required.
+        The dimension of the inputs to the LSTM.
+    output_size : int, required.
+        The dimension of the outputs of the LSTM.
+    direction: str, optional (default = "forward")
```
Comment (matt-gardner) [inline allennlp/modules/augmented_lstm.py]: A string parameter seems kind of loose - maybe just have this be `forward : bool, optional (default=True)`?

File: allennlp/modules/augmented_lstm.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        self.use_highway = use_highway
+
+        if use_highway:
+            self.input_linearity = torch.nn.Linear(input_size, 6 * output_size, bias=True)
+            self.state_linearity = torch.nn.Linear(output_size, 5 * output_size, bias=True)
+        else:
+            self.input_linearity = torch.nn.Linear(input_size, 4 * output_size, bias=True)
+            self.state_linearity = torch.nn.Linear(output_size, 4 * output_size, bias=True)
+
+        self.recurrent_dropout_probability = recurrent_dropout_probability
```
Comment (matt-gardner) [inline allennlp/modules/augmented_lstm.py]: I'd put this up above, next to the other saved parameters.

File: allennlp/modules/augmented_lstm.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    ------
+    output_accumulator : PackedSequence
+        The outputs of the LSTM for each timestep. A tensor of shape
+        (batch_size, max_timesteps, output_size) where for a given batch
+        element, all outputs past the sequence length for that batch are
+        zero tensors.
+    """
+    def __init__(self,
+                 input_size: int,
+                 output_size: int,
```
Comment (matt-gardner) [inline allennlp/modules/augmented_lstm.py]: Do you think it's worth calling this `hidden_size`, to match pytorch's lstm?

File: allennlp/modules/augmented_lstm.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        (batch_size, num_timesteps, output_dimension) representing
+        the outputs of the LSTM per timestep.
+        """
+        if not isinstance(inputs, PackedSequence):
+            raise ConfigurationError('inputs must be PackedSequence but got %s' % (type(inputs)))
+
+        inputs, batch_lengths = pad_packed_sequence(inputs, batch_first=True)
+        batch_size = inputs.size()[0]
+        total_timesteps = inputs.size()[1]
+        output_accumulator = Variable(inputs.data.new().resize_(batch_size,
```
Comment (matt-gardner) [inline allennlp/modules/augmented_lstm.py]: Why do you need to go from `inputs.data.new()` if you're just going to resize and fill?

File: allennlp/modules/augmented_lstm.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+        Returns
+        -------
+        A PackedSequence containing a torch.FloatTensor of shape
+        (batch_size, num_timesteps, output_dimension) representing
+        the outputs of the LSTM per timestep.
+        """
+        if not isinstance(inputs, PackedSequence):
+            raise ConfigurationError('inputs must be PackedSequence but got %s' % (type(inputs)))
+
+        inputs, batch_lengths = pad_packed_sequence(inputs, batch_first=True)
```
Comment (matt-gardner) [inline allennlp/modules/augmented_lstm.py]: What is this line doing?  Also, if you could use a more descriptive name than `inputs`, and have it be different for the two instances here, that'd probably help me know what this is doing better without having to ask.

File: allennlp/modules/augmented_lstm.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        if initial_state is None:
+            full_batch_previous_memory = Variable(torch.zeros([batch_size, self.output_size]))
+            full_batch_previous_state = Variable(torch.zeros([batch_size, self.output_size]))
+        else:
+            full_batch_previous_state = initial_state[0]
+            full_batch_previous_memory = initial_state[1]
+
+        current_length_index = batch_size - 1 if self.direction == "forward" else 0
+        if self.recurrent_dropout_probability > 0.0:
+            dropout_mask = get_dropout_mask(self.recurrent_dropout_probability, [batch_size, self.output_size])
```
Comment (matt-gardner) [inline allennlp/modules/augmented_lstm.py]: Do we need to worry about applying this only at train time?

File: allennlp/modules/augmented_lstm.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+            elif self.direction == "backward":
+                # First conditional: Are we already at the maximum number of elements in the batch?
+                # Second conditional: Does the next shortest sequence beyond the current batch
+                # index require computation use this timestep?
+                while current_length_index < (len(batch_lengths) - 1) and \
+                                batch_lengths[current_length_index + 1] > index:
+                    current_length_index += 1
+
+            # Actually get the slices of the batch which we need for the computation at this timestep.
+            previous_memory = full_batch_previous_memory[0: current_length_index + 1].clone()
```
Comment (matt-gardner) [inline allennlp/modules/augmented_lstm.py]: The space after the `:` here is confusing to me, here and below.  See http://legacy.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements.

File: allennlp/modules/augmented_lstm.py [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+            previous_state = full_batch_previous_state[0: current_length_index + 1].clone()
+            timestep_input = inputs[0: current_length_index + 1, index]
+
+            # Do the projections for all the gates all at once.
```
Comment (matt-gardner) [inline allennlp/modules/augmented_lstm.py]: I think this comment should be earlier, or that there should be a similar comment when you're initializing `self.*_linearity` in the constructor, especially for why the numbers are different in the highway case vs. not.  The way you did this makes a lot of sense (and this is the easiest-to-read LSTM code that I've seen), but it'd be even easier to read with a few more explanatory comments.

File: allennlp/modules/augmented_lstm.py [hunk 13]
```diff
@@ -0,0 +1,171 @@
+from typing import Optional, Tuple
+import torch
+from torch.autograd import Variable
+from torch.nn.utils.rnn import pad_packed_sequence, pack_padded_sequence, PackedSequence
+
+from allennlp.common.tensor import get_dropout_mask
+from allennlp.common.checks import ConfigurationError
+
+
+class AugmentedLstm(torch.nn.Module):
```
Comment (matt-gardner) [inline allennlp/modules/augmented_lstm.py]: Should this be inside `seq2seq_encoders`, and inherit from `Seq2SeqEncoder`?  Or should we wrap this as either a seq2seq or seq2vec encoder, like we do for the pytorch-native ones?

File: allennlp/modules/stacked_alternating_lstm.py [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    A stacked LSTM with LSTM layers which alternate between going forwards over
+    the sequence and going backwards. This implementation is based on the
+    description in `Deep Semantic Role Labelling - What works and what's next
+    <https://homes.cs.washington.edu/~luheng/files/acl2017_hllz.pdf>`_ .
+
+    Parameters
+    ----------
+    input_size : int, required
+        The dimension of the inputs to the LSTM.
+    output_size : int, required
```
Comment (matt-gardner) [inline allennlp/modules/stacked_alternating_lstm.py]: Same question here about whether we should use `hidden_size`.

File: allennlp/modules/stacked_alternating_lstm.py [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+                                  use_highway=use_highway)
+            lstm_input_size = output_size
+            setattr(self, 'layer_{}'.format(layer_index), layer)
+            layers.append(layer)
+        self.lstm_layers = layers
+
+    def forward(self, inputs: PackedSequence):  # pylint: disable=arguments-differ
+        current = inputs
+        for layer in self.lstm_layers:
+            current = layer(current)
```
Comment (matt-gardner) [inline allennlp/modules/stacked_alternating_lstm.py]: I seem to remember Luheng also mentioning dropout between the layers.  I could be remembering wrong, though.

File: allennlp/modules/stacked_alternating_lstm.py [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+                 output_size: int,
+                 num_layers: int,
+                 recurrent_dropout_probability: float = 0.0,
+                 use_highway: bool = True) -> None:
+        super(StackedAlternatingLstm, self).__init__()
+        layers = []
+        lstm_input_size = input_size
+        for layer_index in range(num_layers):
+            direction = "forward" if layer_index % 2 == 0 else "backward"
+            layer = AugmentedLstm(lstm_input_size, output_size, direction,
```
Comment (matt-gardner) [inline allennlp/modules/stacked_alternating_lstm.py]: Would it make sense to have this be configurable?  Just take a `Callable[[], Module]` to construct the layers?  Not sure if it's worth it or not.

File: allennlp/common/tensor.py (hunk 1 above)
Comment (DeNeutoy) [inline allennlp/common/tensor.py]: Duh :man_facepalming:

File: allennlp/modules/augmented_lstm.py (hunk 5 above)
Comment (DeNeutoy) [inline allennlp/modules/augmented_lstm.py]: Yeah - I just thought it wasn't necessarily obvious that `True => forward`, but if it's obvious to you then that's fine.

Comment (matt-gardner) [inline allennlp/modules/augmented_lstm.py]: Well, if the name is `forward`, or `go_forward`, then it should be pretty clear.

File: allennlp/modules/stacked_alternating_lstm.py (hunk 16 above)
Comment (DeNeutoy) [inline allennlp/modules/stacked_alternating_lstm.py]: Not at the moment, because the pytorch native LSTM's do not support going backwards, which is integral to the interleaved thing.

Comment (matt-gardner) [inline allennlp/modules/stacked_alternating_lstm.py]: Oh, right.  They just have bidirectional, not backward.

File: allennlp/modules/stacked_alternating_lstm.py (hunk 15 above)
Comment (DeNeutoy) [inline allennlp/modules/stacked_alternating_lstm.py]: Oh really hmm... I couldn't see it in the paper. I'll check her code.
